src/System.Private.CoreLib/src/System/Type.CoreRT.cs

```diff
@@ -17,7 +19,48 @@ public abstract partial class Type : MemberInfo, IReflect
         public bool IsInterface => (GetAttributeFlagsImpl() & TypeAttributes.ClassSemanticsMask) == TypeAttributes.Interface;
 
         [Intrinsic]
-        public static Type GetTypeFromHandle(RuntimeTypeHandle handle) => RuntimeTypeUnifier.GetTypeForRuntimeTypeHandle(handle);
+        public static Type GetTypeFromHandle(RuntimeTypeHandle handle) => handle.IsNull ? null : GetTypeFromEETypePtr(handle.ToEETypePtr());
+
+        internal static Type GetTypeFromEETypePtr(EETypePtr eeType)
+        {
+            // If we support the writable data section on EETypes, the runtime type associated with the EEType
+            // is cached there. If writable data is not supported, we need to do a lookup in the runtime type
+            // unifier's hash table.
+            if (Internal.Runtime.EEType.SupportsWritableData)
+            {
+                ref GCHandle handle = ref eeType.GetWritableData<GCHandle>();
+                if (handle.IsAllocated)
+                {
+                    return Unsafe.As<Type>(handle.Target);
+                }
+                else
+                {
+                    return GetTypeFromEETypePtrSlow(eeType, ref handle);
+                }
+            }
+            else
+            {
+                return RuntimeTypeUnifier.GetRuntimeTypeForEEType(eeType);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static Type GetTypeFromEETypePtrSlow(EETypePtr eeType, ref GCHandle handle)
+        {
+            // Note: this is bypassing the "fast" unifier cache (based on a simple IntPtr
+            // identity of EEType pointers). There is another unifier behind that cache
+            // that ensures this code is race-free.
+            Type result = RuntimeTypeUnifier.GetRuntimeTypeBypassCache(eeType);
+            GCHandle tempHandle = GCHandle.Alloc(result);
```

Comment:
One thing that I keep coming back to is whether there would be any advantage in making this a weak handle. The fast retrieval path would get slower because now we need to check both whether the handle is allocated and whether the target wasn't collected. But we might be able to reclaim memory. @jkotas do you have an opinion?

Reply:
It is hard to tell whether it would be an improvement.

I think it would be best to match CoreCLR caching strategy. Keep a lightweight System.RuntimeType instance referenced via strong handle. Cache the heavy reflection inside this RuntimeType via a weakhandle. This is the caching strategy that was found to work best for realworld code (iirc, there was investigation on this done for .NET Framework 4.0).

Reply:
I think this is similar to how the reflection stack is structured - the lightweight `RuntimeType` instances hold onto a couple things (e.g. for named types, it holds onto the `RuntimeTypeHandle` (EETypePtr), a lazy cache, the metadata reader, the type definition and type definition handle, and a "namespace chain"). The rest of the "heavy" things are cached in the lazy cache (`TypeComponentsCache`).

The cache currently caches things in a `ConcurrentUnifier`, but maybe it could be a `ConcurrentUnifierW` (the flavor that holds things using a weak reference).

We can look at that at some point.

Reply:
>  a lazy cache, the metadata reader, the type definition and type definition handle, and a "namespace chain"

That is a lot more than in the CoreCLR case. I like the CoreCLR model much better.